### PR TITLE
[dtensor] local_map UX change: keep func signature and be compatible with Tensor input

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126924

**Summary**
This PR has 2 parts of change in `local_map`:

1. regulates the way user can access `DeviceMesh` inside the `func` argument of `local_map`. This means `local_map` will strictly follow the `func` signature without implicitly passing any argument to `func`. If user wants to use `DeviceMesh` inside `func`, this mesh must be explicitly passed to `func` as an argument by user. For example,

```
def user_function(device_mesh, /, *args, **kwargs):
    USER CODE HERE

local_func = local_map(func=user_function, ...)
dtensor_out = local_func(device_mesh, dtensor_input, ...)
```

Before this PR, user code was like:
```
def user_function(device_mesh, /, *args, **kwargs):
    USER CODE HERE

local_func = local_map(func=user_function, ...)
dtensor_out = local_func(dtensor_input, ...)  # local_map passes mesh implicitly for user
```

2. `local_map` now supports mix use of `torch.Tensor` and `DTensor` in argument:

- Pure torch.Tensor case: no `DTensor` argument is passed in, all tensor arguments are `torch.Tensor`. Bypass the `in_placements` check and unwrapping steps. The output will not be wrapped into `DTensor` but directly returned.
- Pure DTensor case: no `torch.Tensor` argument is passed in, all tensor arguments are `DTensor`. This follows the default rule: `in_placements` check, unwrapping arguments, pass into `func`, wrapping the `torch.Tensor` output into `DTensor` if the `out_placements` is not `None`.
- Mix of the above two: some arguments are `torch.Tensor` while some are `DTensor`. Only perform `in_placements` check and unwrapping on `DTensor` arguments. For output processing, it's the same as Pure DTensor case.

**Test**
`pytest test/distributed/_tensor/experimental/test_local_map.py`

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @msaroufim

### CODE REVIEW

**Review by wanchaol — CHANGES_REQUESTED** [review]
The signature change lgtm! I would like to see bunch of things happen in this PR together:

1. We should force the check of ACT when we call `to_local` in the local_map, and wait on ACT inside `local_map` itself instead of user code
2. We should regulate the in-placements and out-placements type. for in_placements and out placements, we should: 1) make the in_placements force to pass a tuple of placements, not accepting list, so it should be (arg_1_list_placements, arg_2_list_placements, ...) 2. similar restriction to put into out_placements. For out placements, we should force user to pass a tuple of output_placements, instead of being flexible and accept both list/tuples. This is so that we avoid confusing users about the UX when using it

File: torch/distributed/_tensor/experimental/local_map.py [hunk 1]
```diff
@@ -139,44 +157,62 @@ def wrapped(*args, **kwargs):
...
+            # process output
+            flat_out, out_spec = pytree.tree_flatten(out)
 
-                flat_dist_out.append(
-                    DTensor.from_local(out, device_mesh, spec, run_check=False)
+            flat_dist_out = []
+            for idx, out in enumerate(flat_out):
+                spec = (
+                    out_placements[idx]
+                    if isinstance(out_placements, tuple)
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/local_map.py:195]: we should have `out_placements = tuple(out_placements) if not isinstance(out_placements, tuple)` first, and then in this for loop we can simply `for out, out_placement in zip(flat_out, out_placements)`?

File: test/distributed/_tensor/experimental/test_local_map.py [hunk 2]
```diff
@@ -99,14 +109,40 @@ def test_local_map_out_placements(self):
...
         self.assertTrue(not (X.equal(Y)))
 
+        # Test 2: directly return out if no argument is DTensor
+        # matmul in DDP
+        replicate = [Replicate()]
+        X = torch.randn(
+            4 // self.world_size, 4, device=self.device_type, requires_grad=False
+        )
+        W = torch.randn(4, 4, device=self.device_type, requires_grad=False)
+        local_mm_all_gather_forward = local_map(
```
Comment (wanchaol) [inline test/distributed/_tensor/experimental/test_local_map.py:130]: can we add a explicit test case to test the `functools.partial` decorator works?

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for improving this! have some comments inlined.

File: test/distributed/_tensor/experimental/test_local_map.py (hunk 2 above)
Comment (XilunWu) [reply test/distributed/_tensor/experimental/test_local_map.py:130]: no problem, i'll add `test_local_map_decorator` in a follow up PR.

File: torch/distributed/_tensor/experimental/local_map.py (hunk 1 above)
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/local_map.py:195]: That sounds better, let me address it in a follow up along with adding the test.

Comment (XilunWu) [reply torch/distributed/_tensor/experimental/local_map.py:195]: comment addressed in #127752

**Pull request closed without merging**